content/posts/2025-02-05-perf-improvements.md

```diff
@@ -0,0 +1,237 @@
+
+## Next
+
+Better performance is a never ending battle. In NetObserv, it is critical to run our eBPF probes with the minimal possible impacts on the monitored workloads and on the cluster overall.
+
+### Back to the ringbuffer
+
+As mentioned above, there are some ideas to improve further the data processing between kernel and user spaces. One of them consists in re-evaluating the relevance of a [ringbuffer](https://docs.ebpf.io/linux/map-type/BPF_MAP_TYPE_RINGBUF/).
+
+[Previously](https://opensource.com/article/22/8/ebpf-network-observability-cloud) we found that a hash map is more relevant than a ringbuffer for forwarding flows to the user space. This is mostly because the hash map allows in-kernel aggregation, thus reducing the size of the data to copy, whereas the ringbuffer is for a stateless usage, transferring data as soon as it comes, thus with the risk to copy some duplicated parts again and again.
+
+However, now that we have split the data into two maps, while the above certainly remains true for the main flows map, we can't say the same for the enrichment map. Enriched data is less subject to in-kernel aggregation, and the enrichment specialized hooks are less often triggered. So it's surely worth it to re-evaluate in that context.
+
+### Even more de-duplication
+
+We've seen that doing de-duplication in the kernel already helped a lot. There's more that we can do, but to keep nothing from you, it's a bit tricky.
+
+Modern networking is complex and multi-layered. There are more layers than what we currently de-duplicate. In OVN-Kubernetes for instance, a [GENEVE encapsulation](https://www.redhat.com/en/blog/what-geneve) is used for sending packets between nodes. When it happens, NetObserv sees different things. There's the pod-to-pod traffic, for instance, that is identified with the corresponding Pod IPs. That's one flow. And at the same time there's the node-to-node traffic, which is seen from another interface like `br-ex` and identified with the source and destination Node IPs. Two flows, same packet (encapsulated or not).
+
+On top of that, there are also the Kubernetes Services that show up with different IPs — so, different flows: remember that a flow is defined by its IPs — even if that's for the same packet. This is resolved through NAT.
+
+The question remains open whether we want to de-duplicate that or not. It may impact not only the eBPF agent, but also how things are displayed in the Console plugin, as we'll need to offer a more layered view of the flows. So it is also a UX challenge. Note that very recently, we also started to cover Service NAT by a different means, through [conntrack](https://netobserv.io/posts/enhancing-netobserv-for-kubernetes-service-flows-using-ebpf/). But that doesn't actually *de-duplicate*, it's an enrichment.
```

Comment:
nit: not only service NAT, its more of NAT in general service or egressIP or for any other reasons

Reply:
ok thanks I'll update